lib/mongo/socket.rb

```diff
@@ -172,14 +172,16 @@ def read_from_socket(length)
       # Block on data to read for timeout seconds
       # using the suggested implementation of http://www.ruby-doc.org/core-2.1.3/Kernel.html#method-i-select
       # to work with SSL connections and some OSs like Ubuntu, CentOS and Red Hat
-      time_left = timeout.to_f
       data = String.new
+      deadline = (Time.now + timeout) if timeout
       begin
-        while data.length < length
+        while (data.length < length)
+          if deadline
```

Comment:
Seems more logical to move this block into the rescue since you should always be reading at least once.